### PR TITLE
fix: update CUGA component to use SDK with streaming and fix Langfuse v3 compatibility

### PR DESCRIPTION
**Pull request by haroldship**
### CUGA Component — SDK Integration with Streaming
- Switch from `invoke()` to `stream()` for real-time UI feedback during agent execution
- Parse LangGraph-style `(node_name_tuple, state_dict)` stream format from CUGA SDK
- Extract `final_answer` from nested CUGA state, with fallback to last AI chat message
- Emit `on_chain_start`, `on_tool_start`, `on_tool_end`, `on_chain_end` events for UI visibility

### CUGA Component — Custom Tool Provider
- Add `LangflowToolsProvider` implementing CUGA SDK's `ToolProviderInterface`
- Three grouping strategies: by `server_name` metadata → by common prefix → default app
- App descriptions follow SDK spec: tool name concatenation for server/prefix apps, name + description[:300] for default app
- Enables CUGA's tool shortlisting (`find_tools`) for large tool sets

### CUGA Component — New Inputs
- Add `shortlisting_tool_threshold` input (default 35, advanced) — enables `find_tools` when total tools exceed threshold
- Remove `web_apps` input (no longer needed with custom tool provider)
- Add `decomposition_strategy` input (flexible/exact, advanced)

### Langfuse v3 API Compatibility
- Migrate from v2 (`trace()`, `span()`) to v3 API (`start_span()`, `update_trace()`, `end()`)
- Use `Langfuse.create_trace_id(seed=...)` for deterministic trace IDs
- Use `auth_check()` for health checks
- Support `parent_span_id` in `get_langchain_callback()` for proper nesting
- Integrate tracing callbacks with CUGA agent execution

### CodeRabbit Review Fixes
- Add stream fallback: use last AI chat message when `final_answer` is empty
- Add `self.apps.clear()` / `self.tools_by_app.clear()` for idempotent `initialize()`
- Add explicit `auth_check` mocking in Langfuse tests
- Remove redundant `@pytest.mark.asyncio` decorators (pyproject.toml has `asyncio_mode = "auto"`)

## Files Changed
| File | Changes |
|------|---------|
| `src/lfx/src/lfx/components/cuga/cuga_agent.py` | SDK integration, streaming, `LangflowToolsProvider`, new inputs |
| `src/backend/base/langflow/services/tracing/langfuse.py` | Langfuse v3 API migration |
| `src/backend/tests/unit/components/models_and_agents/test_cuga_agent.py` | Tests for provider grouping, streaming, component fields |
| `src/backend/tests/unit/services/tracing/test_langfuse.py` | Tests for Langfuse v3 API (new file) |
| `src/backend/tests/unit/services/tracing/test_langfuse_v3_compatibility.py` | Additional v3 compatibility tests (from feat/langchain-1.0) |

## Test plan
- [x] Unit tests pass: `test_cuga_agent.py` (21 passed, 8 skipped)
- [x] Unit tests pass: `test_langfuse.py` (8 passed)
- [x] `make format_backend` passes
- [x] Manual browser testing: CUGA with MCP tools (email, file_system)
- [x] Verified tool grouping by server_name in UI
- [x] Verified streaming events appear during execution
- [x] Verified tool shortlisting with `find_tools` at threshold=3
- [x] Verified agent discovers and uses tools via `find_tools` (file read test)
- [x] Verified full tool visibility at threshold=35 (no shortlisting)

### CODE REVIEW

File: src/lfx/src/lfx/components/cuga/cuga_agent.py
Comment (sami-marreed) [inline src/lfx/src/lfx/components/cuga/cuga_agent.py:271]: Remove

Comment (sami-marreed) [reply src/lfx/src/lfx/components/cuga/cuga_agent.py:271]: Add the tool shortlisting threshold ( when to activate find tools tool ) based on a threshold
```
shortlisting_tool_threshold = 35  # Enable find_tools if total number of tools exceeds this threshold
```

File: src/lfx/src/lfx/components/cuga/cuga_agent.py [hunk 1]
```diff
@@ -210,51 +210,48 @@ async def call_agent(
...
-                await cuga_agent.initialize_appworld_env()
+
+            # Get Langflow tracing callbacks (includes Langfuse, LangSmith, etc.)
+            langchain_callbacks = self.get_langchain_callbacks()
+            logger.debug(f"[CUGA] Got {len(langchain_callbacks)} tracing callbacks: {[type(cb).__name__ for cb in langchain_callbacks]}")
+
+            # Create the CUGA SDK agent with tools directly (no tracking)
+            logger.debug(f"[CUGA] Creating CugaAgent with {len(tools)} tools")
+            cuga_agent = CugaAgent(
+                tools=tools,
```
Comment (sami-marreed) [inline src/lfx/src/lfx/components/cuga/cuga_agent.py]: Consider switching to custom tool provider /cuga-agent/src/cuga/backend/cuga_graph/nodes/cuga_lite/tool_provider_interface.py `CugaToolsProvider`, with get_tools, get_apps, get_all_tools, to create apps do it based on three strategies there's already old code in cuga how to do this under src/cuga/backend/activity_tracker/tracker.py `set_tools` function
1. if tool.metadata.get("server_name", None)( source mcp server name ) is == across multiple tools, and provide the desciprtion of the application as concatenation of tool names
2. by common prefix for example gmail tools they are not mcp server, but they have gmail_<suffix>
3. add them as `default` app with description and as concatenation of tool name and its description[:300]

File: src/lfx/src/lfx/components/cuga/cuga_agent.py
Comment (sami-marreed) [inline src/lfx/src/lfx/components/cuga/cuga_agent.py]: Maybe we should remove this?

Comment (haroldship) [reply src/lfx/src/lfx/components/cuga/cuga_agent.py:271]: done

File: src/lfx/src/lfx/components/cuga/cuga_agent.py (hunk 1 above)
Comment (haroldship) [reply src/lfx/src/lfx/components/cuga/cuga_agent.py]: done

File: src/lfx/src/lfx/components/cuga/cuga_agent.py
Comment (haroldship) [reply src/lfx/src/lfx/components/cuga/cuga_agent.py]: done

**Comment by haroldship** [comment]
@coderabbitai review

**Comment by haroldship** [comment]
@ogabrielluiz this PR will require changes to the documentation because we removed The web_apps and added the shortlisting_tool_threshold inputs

**Comment by ogabrielluiz** [comment]
> @ogabrielluiz this PR will require changes to the documentation because we removed The web_apps and added the shortlisting_tool_threshold inputs

@mendonk

File: src/lfx/src/lfx/components/cuga/cuga_agent.py [hunk 2]
```diff
@@ -269,58 +417,110 @@ async def call_agent(
...
+                end_event = {
+                    "event": "on_chain_end",
+                    "run_id": str(uuid.uuid4()),
+                    "name": "CugaAgent",
+                    "data": {"output": AgentFinish(return_values={"output": final_answer}, log="")},
+                }
+                logger.info(
+                    f"[CUGA] Yielding chain_end event with answer: {final_answer[:100] if final_answer else 'None'}..."
+                )
+                yield end_event
```
[the remark this answers is not in the document]
Comment (haroldship) [reply src/lfx/src/lfx/components/cuga/cuga_agent.py:522]: I've fixed this
